simulation/traffic_simulator/include/traffic_simulator/api/api.hpp

```diff
@@ -157,13 +165,6 @@ class API
     return register_to_entity_manager() and register_to_environment_simulator();
   }
 
-  geometry_msgs::msg::Pose toMapPose(const geometry_msgs::msg::Pose & pose) { return pose; }
-
-  geometry_msgs::msg::Pose toMapPose(const traffic_simulator_msgs::msg::LaneletPose & pose)
-  {
-    return entity_manager_ptr_->getHdmapUtils()->toMapPose(pose).pose;
-  }
-
   template <typename Pose>
   auto spawn(
     const std::string & name, const Pose & pose,
```

Comment:
Member functions of the API class should not accept LanaletPose to prevent incorrect values from being entered by the user, but instead accept CanonicalizedLaneletPose.

Reply:
@hakuturu583 
Please take a look at that, for the _spawn_ operation, canonicalization takes place in the `EntityManager`, more precisely here:

https://github.com/tier4/scenario_simulator_v2/blob/90b486d9b38739d1b9a1abf2d150eb5df081fc67/simulation/traffic_simulator/include/traffic_simulator/entity/entity_manager.hpp#L473-L488

It is worth to note that for canonicalization (`toCanonicalizedLaneletPose`), sometimes the bounding box and matching distance are used. These are not available for an `Entity` that is not spawned yet (it is not in the `EntityManager`). The only thing that can be done in such a situation is to perform a simple canonicalization as in this line.

https://github.com/tier4/scenario_simulator_v2/blob/90b486d9b38739d1b9a1abf2d150eb5df081fc67/simulation/traffic_simulator/include/traffic_simulator/entity/entity_manager.hpp#L482

So I believe that the safeguard you mention is already implemented, but the responsibility for this is on the side of the `EntityManager`.

It was developed in such a way that the user of the `API` did not have to keep in mind to perform canonicalization, what's more, to ensure that it is always performed in the same way - no matter if the api is used by `SimulatorCore` or by `CppMockScenarioz`. I believe that, during `spawn()` in `CppMockScenarios` - which is called quite often, it is convenient to simply specify `LaneletPose`, and the responsibility for proper conversion or validation should be strictly inside `traffic_simulator` (since `CanonicalizedLaneletPose` is an "internal type" rather than created to be used for communication - especially due to various canonicalization methods - with bounding box, with different matching distance etc.).

To summarize:
I believe that `spawnEntity()` should use `msg::Pose` and `msg::LaneletPose`. `CanonicalizedLaneletPose` is currently allowed due to the "legacy" side of `SimulatorCore` - it still uses `CanonicalizedLaneletPose`. I believe that the usage of `CanonicalizedLaneletPose` type in `openscenario_interpreter` and `SimulatorCore` can be an unclean and troublesome approach. I believe that we should take care that the conversion between `msg::LaneletPose` and `CanonicalziedLaneletPose` is done in strictly fixed parts and uses as uniform a method as possible (e.g. always the same `matching_distance`). 

Please take a look at [this PR](https://github.com/tier4/scenario_simulator_v2/pull/1252/files), where I plan to introduce more changes related `toCanonicalizeLaneletPose()`.


For this reasons, I believe that passing `msg::LaneletPose` in `API::spawn()` and ensuring that proper canonicalization is performed inside `EntityManager` can help in the future take it in a better direction.

Reply:
> CanonicalizedLaneletPose is currently allowed due to the "legacy" side of SimulatorCore - it still uses CanonicalizedLaneletPose. I believe that the usage of CanonicalizedLaneletPose type in openscenario_interpreter and SimulatorCore can be an unclean and troublesome approach.

That is not true.
CanonicalizedLanaletPose is designed to prevent incorrect values from being entered using the C++ type system and should be used inside scenario_simulator_v2 except when pub/sub is used.

Reply:
It also confuses users because functions that use CanonicalizedLanaletPose and traffic_simulator_msgs::msg::LanaletPose inside the API class are in disorder.

In order to prevent incorrect values from being entered and to ensure consistency of API functions, the API class argument type should be CanonicalizedLanaletPose instead of traffic_simulator_msgs::msg::LanaletPose.

Reply:
@hakuturu583 
I understand that `CanonicalizedLaneletPose` was introduced to prevent entering incorrect values in `msg::LaneletPose`. I just want to point out that in ss2 code there was a widespread problem with using different arguments  when `CanonicalizedLaneletPose` is created (generally when `msg::Pose` is used).

But you are absolutely right that in the case of conversion `msg::LaneletPose`<->`CanonicalizedLaneletPose` the results will be the same, because in that case `pose::canonicalize()` is used for the conversion, or more precisely simply the [`CanonicalizedLaneletPose()`](https://github.com/tier4/scenario_simulator_v2/blob/ref/RJD-1053-set-update-canonicalized-entity-status/simulation/traffic_simulator/src/data_type/lanelet_pose.cpp) constructor is called, where the `HdMapUtils::canonicalizeLaneletPose` is called:

https://github.com/tier4/scenario_simulator_v2/blob/bb2fe152b2b2be7d926d88dcdb611a460c7cd456/simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp#L157-L179

So conversion `msg::LaneletPose`<->`CanonicalizedLaneletPose` can be successfully done even in `SimulatorCore` or `CppMockScenarios`, but I don't really understand the advantage of "extracting" this conversion outside `traffic_simulator`. 

Could you please let me know if there is any reason why `CanonicalizedLaneletPose` is used for `SimulatorCore`<->`API` communication instead of `msg::LaneletPose`? :bow:  This conversion requires access to lanelet2 (`HdMapUtils`). This introduces a greater dependency between `openscenario_interpreter` and `traffic_simulator`, which can probably be fully avoided by using `msg::LaneletPose` while keeping the code clear. 

Unfortunately, all my knowledge is based on code analysis, and there are certainly arrangements (including those for the future) that I'm unfamiliar with.

Of course, I understand situations such as calling [`entity->getCanonicalizedLaneletPose()`](https://github.com/tier4/scenario_simulator_v2/blob/bb2fe152b2b2be7d926d88dcdb611a460c7cd456/openscenario/openscenario_interpreter/include/openscenario_interpreter/simulator_core.hpp#L151) in `SimulatorCore` and its use, but passing `CanonicalizedLaneletPose` from `openscenario_interpreter` to `traffic_simulator` (e.g., to [set a request](https://github.com/tier4/scenario_simulator_v2/blob/bb2fe152b2b2be7d926d88dcdb611a460c7cd456/simulation/traffic_simulator/src/entity/vehicle_entity.cpp#L183-L203) or spawn) is unclear to me, since `pose::canonicalize()` can be done in `traffic_simulator`, which will help achieve fully my intentions I'm trying to outline below.

From my point of view, it would be good for canonicalization (both `pose::canonicalization()` and `pose::toCanonicalizedLaneletPose()`) to be performed possibly in the same parts of the code. I would also like to add, that with this PR and the [next one](https://github.com/tier4/scenario_simulator_v2/pull/1252/files) - I want to reduce confusion with the conversion of `msg::Pose`/`msg::LaneletPose` to `CanonicalizedLaneletPose`.

I would also like to highlight the issue with multiplicity of canonicalizations. The issue is significant due to the widespread use of `EntityStatus`, where `msg::LaneletPose` occurs, and every time `EntityStatus` is passed and used, another `pose::canonicalize()` must be performed. Of course, the use of `EntityStatus` is unavoidable, but I believe we can make sure that canonicalization is performed a much smaller number of times, in fixed parts of the code and with greater clarity - possibly in the same parts of the code.

For these reasons I've attempted (possibly incorrectly) to unify this and ensure that the `msg::Pose`/`msg::LaneletPose`<->`CanonicalizedLaneletPose` conversion is performed using the attributes of the specific `Entity` - that is, it is performed "close" to the `Entity` - as here:

https://github.com/tier4/scenario_simulator_v2/blob/b9a8c6eb397236a3e06ed9834b8b747f2579935e/simulation/traffic_simulator/src/api/api.cpp#L121-L142

In the [next PR](https://github.com/tier4/scenario_simulator_v2/pull/1252/files), I would like to move `toCanonicalizedLaneletPose()` even more "close" to the Entity (inside `CanonicalizedEntityStatus::set()`:

https://github.com/tier4/scenario_simulator_v2/blob/bb2fe152b2b2be7d926d88dcdb611a460c7cd456/simulation/traffic_simulator/src/entity/entity_base.cpp#L520-L523

https://github.com/tier4/scenario_simulator_v2/blob/bb2fe152b2b2be7d926d88dcdb611a460c7cd456/simulation/traffic_simulator/src/data_type/entity_status.cpp#L66-L84

Reply:
> Could you please let me know if there is any reason why CanonicalizedLaneletPose is used for SimulatorCore<->API communication instead of msg::LaneletPose? 🙇 This conversion requires access to lanelet2 (HdMapUtils). This introduces a greater dependency between openscenario_interpreter and traffic_simulator, which can probably be fully avoided by using msg::LaneletPose while keeping the code clear.

I believe that API arguments need to be consistent.
The API should take either LaneletPose or CanonicalizedLaneletPose type arguments, and it is not a good design to accept both for each function.

In the current implementation, only the API::spawn function accepts CanonicalizedLaneletPose, which is inconsistent with other functions.
However, changing all CanonicalizedLaneletPose in API class arguments to LaneletPose would involve a major change, so a separate Pull Request is necessary.
Therefore, in this PR, please revert back to this API::spawn function argument type change.

Reply:
@hakuturu583 
Certainly, [done](https://github.com/tier4/scenario_simulator_v2/pull/1266/commits/6d42b00554c71f3c31d8acfaf02b068c446fee5f) ([cpp_mock_scenarios adjustments](https://github.com/tier4/scenario_simulator_v2/pull/1266/commits/6a74a495a028fbb9f711efc91080efa45d98d8bb)). :bow: 

Please let me know if you see needs for any more changes. 
And referring to [this comment](https://github.com/tier4/scenario_simulator_v2/pull/1266#issuecomment-2167731500), please let me know if you recommend rerunning the evaluator.